src/buildercore/terraform.py

```diff
@@ -1229,6 +1229,46 @@ def _render_eks_master_security_group(context, template):
         'tags': security_group_tags,
     })
 
+def _render_eks_addon(context, template, name, addon_name, version, configuration_values=False, resolve_conflicts='OVERWRITE', service_account_role_arn=False):
+    if (version == 'latest'):
+        template.populate_data(
+            'aws_eks_addon_version',
+            'eks_addon_%s' % name,
+            block={
+                'addon_name': addon_name,
+                'kubernetes_version': '${data.aws_eks_cluster.main.version}',
+                'most_recent': True,
+            }
+        )
+
+    resource_block = {
+        'cluster_name': '${data.aws_eks_cluster.main.id}',
+        'addon_name': addon_name,
+        'addon_version': version if version != 'latest' else '${data.aws_eks_addon_version.eks_addon_%s.version}' % name,
+        'tags': aws.generic_tags(context),
+        'resolve_conflicts': resolve_conflicts,
+    }
+
+    if (configuration_values != False):
+        resource_block['configuration_values'] = configuration_values
+
+    if (service_account_role_arn != False):
+        resource_block['service_account_role_arn'] = service_account_role_arn
+
+    template.populate_resource(
+        'aws_eks_addon',
+        'eks_addon_%s' % name,
+        block=resource_block,
+    )
+
+def _render_eks_addons(context, template):
+    if (lookup(context, 'eks.addons.kube-proxy')):
+        addon_version = lookup(context, 'eks.addons.kube-proxy.version', 'latest')
+        _render_eks_addon(context, template, 'kube_proxy', 'kube-proxy', addon_version)
+    if (lookup(context, 'eks.addons.coredns')):
```

Comment:
This is beginning to look repetitive, and you're hardcoding specific addons. Perhaps:

```python
for addon_name, addon_context in context["eks"]["addons"].items():
    _render_eks_addon(context, template, addon_name, addon_context)
```

And change the signature of `_render_eks_addon` to accept the 'addon context'.

The function `_render_eks_addon` is also introducing values that should be handled during context building:
* `configuration_values=False`
* `resolve_conflicts='OVERWRITE'`
* `service_account_role_arn=False`

Reply:
Yup I did hardcode them intentionally - these are simple addons, but the next ones require coordination between IAM, cluster and addon resources. My intention was to build that knowledge into code, not config - so when `addon.vpc-cni` is configured, then verify the IAM OIDC is setup, add additional IAM resources needed and set `service_account_role_arn`. So the next if blocks would have had multiple `_render*` function calls. Otherwise, I'd have to assume the required parts are configured and leave it as an exercise to the user to figure out which config is needed to work together, potentially leaving a cluster in a broken state if they don't. 

Assuming you agree fewer foot-guns are better - I think then I will probably need to move some of the current IAM logic up the stack from `terraform.py` into `cfngen.py` to to pull that off, but it sounds like that logic should always have been there. 

While I'm in this head space - some of these configurations have been there to aid migration (i.e. do it on test first, iterate until it works, then roll out to prod), but ideally we would always configure them now the migration is done to remove the number of moving parts. An example is this one - if I can assume our configuration of EKS always enables the IAM-OIDC, then I don't need to verify that is is enabled (so at least 1 less condition and 1 less verification, possibly simpler tests, etc). However, it is still valid to have a cluster without this IAM integration, I just don't see us/me ever using it. Is that your preference from a builder perspective - fewer, but more integrated configurations? or would it be better to keep configuration around?

Reply:
When dealing with cloudformation/terraform I suggest taking the code from one stable, tested, coherent working version to another. If this code works for your current usecase, merge it in [you just did ;)]. 

If you *know* configuration will be short-lived and won't be reused or extended in the future, feel free to hardcode it in, just remember to also remove it when the transition is done.